larpandoradlcontent/LArTwoDReco/DLThreeDClusterSplittingAlgorithm.cc

```diff
@@ -0,0 +1,1350 @@
+
+        ClusterHit clusterHit(pCaloHit, lBinIndex, thisHitT);
+
+        const auto iter(std::find(clusterPath.begin(), clusterPath.end(), clusterHit));
+
+        if (iter == clusterPath.end())
+        {
+            clusterPath.emplace_back(clusterHit);
+        }
+        else if (thisHitT < iter->m_t)
+        {
+            *iter = clusterHit;
+        }
+    }
+
+    // Order low->high l
+    std::sort(clusterPath.begin(), clusterPath.end(), [](const ClusterHit &lhs, const ClusterHit &rhs) { return lhs.m_l < rhs.m_l; });
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+void DLThreeDClusterSplittingAlgorithm::InitialiseFeatures(Features &features) const
+{
+    features.insert(std::make_pair("Longitudinal", Feature((0.0), 1.0, std::vector<float>())));
+    features.insert(std::make_pair("Transverse", Feature((-0.01), 2.57, std::vector<float>())));
+    features.insert(std::make_pair("Energy", Feature(0.49, 0.28, std::vector<float>())));
+    features.insert(std::make_pair("HitWidth", Feature(0.56, 0.16, std::vector<float>())));
+    features.insert(std::make_pair("Theta", Feature((-0.01), 0.11, std::vector<float>())));
+    features.insert(std::make_pair("SecVertex", Feature(93.19, 103.39, std::vector<float>())));
+    features.insert(std::make_pair("GapSep", Feature(0.0, 1.0, std::vector<float>())));
+    features.insert(std::make_pair("EventHitSep", Feature(5.00, 5.83, std::vector<float>())));
+    features.insert(std::make_pair("ClusterHitSep", Feature(0.54, 0.13, std::vector<float>())));
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+void DLThreeDClusterSplittingAlgorithm::FillFeatures(const ClusterPath &clusterPath, const TwoDSlidingFitResult &clusterFit, Features &features,
+    HitKDTree2D &kdTree) const
+{
+    // Get path hits, and their total energy
+    CaloHitList clusterPathHits; float totalEnergy(0.f);
+    for (const auto &entry : clusterPath)
+    {
+        const CaloHit *const pPathCaloHit(entry.m_pHit);
+        totalEnergy += pPathCaloHit->GetElectromagneticEnergy();
+        clusterPathHits.push_back(pPathCaloHit);
+    }
+
+    // Initialise KalmanFilter
+    KalmanFilter2D kalmanFilter2D(this->InitialiseKalmanFilter(clusterPath));
+
+    // To reduce computations and loops, get filtered secVertex list in view
+    CartesianPointVector viewSecVtx;
+    this->GetFilteredViewSecVertices(clusterFit, viewSecVtx);
+
+    // Fill features
+    bool processedFirst(false); float cumulativeEnergy(0.f);
+    for (unsigned int i = 0; i < clusterPath.size(); ++i)
+    {
+        const ClusterHit &clusterHit(clusterPath.at(i));
+
+        if (!processedFirst)
+        {
+            features.at("Theta").m_sequence.push_back(-4.f);
+            processedFirst = true;
```

Comment:
1) A magic number? 2) What's this actually doing, presumably accommodating an ill-defined opening angle prior to having two hits with which to compute one?